src/sabre_swap/mod.rs

```diff
@@ -410,99 +324,216 @@ fn swap_map_trial(
-    // Run in parallel only if we're not already in a multiprocessing context
-    // unless force threads is set.
-    let candidate_swaps = obtain_swaps(layer, neighbor_table, layout);
-    let mut min_score = f64::MAX;
-    let mut best_swaps: Vec<[usize; 2]> = Vec::new();
-    for swap_qubits in candidate_swaps {
-        layout.swap_logical(swap_qubits[0], swap_qubits[1]);
-        let score = score_heuristic(
-            heuristic,
-            layer,
-            extended_set,
-            layout,
-            &swap_qubits,
-            dist,
-            qubits_decay,
-        );
-        if score < min_score {
-            min_score = score;
-            best_swaps.clear();
-            best_swaps.push(swap_qubits);
-        } else if score == min_score {
-            best_swaps.push(swap_qubits);
+/// Update the system state as the given `nodes` are added to the routing order, preceded by the
+/// given `swaps`.  This involves updating the output values `gate_order` and `out_map`, but also
+/// the tracking objects `front_layer`, `extended_set` and `required_predecessors` by removing the
+/// routed nodes and adding any now-reachable ones.
+fn update_route(
+    nodes: &[NodeIndex],
+    swaps: Vec<[usize; 2]>,
+    dag: &SabreDAG,
+    layout: &NLayout,
+    coupling: &DiGraph<(), ()>,
+    gate_order: &mut Vec<usize>,
+    out_map: &mut HashMap<usize, Vec<[usize; 2]>>,
+    front_layer: &mut FrontLayer,
+    extended_set: &mut ExtendedSet,
+    required_predecessors: &mut [u32],
+) {
+    // First node gets the swaps attached.  We don't add to the `gate_order` here because
+    // `route_reachable_nodes` is responsible for that part.
+    let py_node = dag.dag[nodes[0]].0;
+    out_map.insert(py_node, swaps);
+    for node in nodes {
+        front_layer.remove(node);
+    }
+    route_reachable_nodes(
+        nodes,
+        dag,
+        layout,
+        coupling,
+        gate_order,
+        front_layer,
+        required_predecessors,
+    );
+    // Ideally we'd know how to mutate the extended set directly, but since its limited size ties
+    // its construction strongly to the iteration order through the front layer, it's not easy to
+    // do better than just emptying it and rebuilding.
+    extended_set.clear();
+    populate_extended_set(extended_set, dag, front_layer, required_predecessors);
+}
+
+/// Search forwards in the `dag` from all the nodes in `to_visit`, adding them to the `gate_order`
+/// or the current `front_layer` as appropriate, and continue inspecting gates until there is
+/// nothing further with no required predecessors.
+///
+/// The nodes in `to_visit` should all already have no further required predecessors.
+fn route_reachable_nodes(
+    to_visit: &[NodeIndex],
+    dag: &SabreDAG,
+    layout: &NLayout,
+    coupling: &DiGraph<(), ()>,
+    gate_order: &mut Vec<usize>,
+    front_layer: &mut FrontLayer,
+    required_predecessors: &mut [u32],
+) {
+    let mut to_visit = to_visit.to_vec();
```

Comment:
I'm wondering if we can get away without copying here somehow. I haven't thought it through in depth but my gut instinct was we could potentially write this function without having to append to the input node list like this. But I'm probably just overlooking something.

Not that this really matters, and I doubt this has any real impact on performance, especially since the routable nodes list is never going to be that long.

Reply:
even in the worst case, we could just write a small class that borrows the input `Vec`, and only allocates new heap space on `push`, and redirects the indexing to the right places in it.  I didn't want to get too deep into these optimisations quite yet for now.